File: clients/go/node/keymgr.go
package main

import (
	"encoding/hex"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"strings"

	"rubin.dev/node/crypto"
)

// Q-127: key lifecycle tooling (non-consensus).
//
// This implements a dev keystore format backed by AES-256-KW:
// - strict mode: requires wolfcrypt shim keywrap symbols
// - non-strict mode: allows software AES-KW fallback for dev/test environments

type keyWrapProvider interface {
	HasKeyManagement() bool
	KeyWrap(kek, keyIn []byte) ([]byte, error)
	KeyUnwrap(kek, wrapped []byte) ([]byte, error)
}

type KeyStoreV1 struct {
	Version      string `json:"version"` // "RBKSv1"
	SuiteID      uint8  `json:"suite_id"`
	PubkeyHex    string `json:"pubkey_hex"`
	KeyIDHex     string `json:"key_id_hex"`
	WrapAlg      string `json:"wrap_alg"` // "AES-256-KW"
	WrappedSKHex string `json:"wrapped_sk_hex"`
}

func wolfcryptStrict() bool {
	v := os.Getenv("RUBIN_WOLFCRYPT_STRICT")
	return v == "1" || strings.EqualFold(v, "true")
}

func mustLen(b []byte, n int, name string) error {
	if len(b) != n {
		return fmt.Errorf("%s must be %d bytes (got %d)", name, n, len(b))
	}
	return nil
}

func loadKeyWrapper(strict bool) (keyWrapProvider, func(), error) {
	// In strict mode: require shim keywrap symbols.
	// In non-strict mode: allow software AES-KW fallback for dev/test.
	p, cleanup, err := loadCryptoProvider()
	if err != nil {
		return nil, func() {}, err
	}
	if km, ok := p.(keyWrapProvider); ok && km.HasKeyManagement() {
		return km, cleanup, nil
	}
	cleanup()
	if strict {
		return nil, func() {}, fmt.Errorf("keymgr requires wolfcrypt shim keywrap symbols in strict mode")
	}
	return nil, func() {}, nil
}

func keywrap(strict bool, kek, keyIn []byte, km keyWrapProvider) ([]byte, error) {
	if km != nil {
		return km.KeyWrap(kek, keyIn)
	}
	return crypto.AESKeyWrapRFC3394(kek, keyIn)
}

func keyunwrap(strict bool, kek, wrapped []byte, km keyWrapProvider) ([]byte, error) {
	if km != nil {
		return km.KeyUnwrap(kek, wrapped)
	}
	if strict {
		return nil, fmt.Errorf("keyunwrap requires shim in strict mode")
	}
	return crypto.AESKeyUnwrapRFC3394(kek, wrapped)
}

func cmdKeymgrExportWrapped(argv []string) error {
	fs := flag.NewFlagSet("keymgr export-wrapped", flag.ExitOnError)
	out := fs.String("out", "", "output keystore json path")
	suiteID := fs.Uint("suite-id", 0, "suite_id (0x01 ML-DSA-87; 0x02 SLH-DSA reserved)")
	pubkeyHex := fs.String("pubkey-hex", "", "pubkey bytes (hex)")
	skHex := fs.String("sk-hex", "", "secret key bytes (hex) to wrap (dev only; do not use for FIPS claims)")
	kekHex := fs.String("kek-hex", "", "AES-256 KEK (32 bytes hex)")
	_ = fs.Parse(argv)
	if *out == "" || *pubkeyHex == "" || *skHex == "" || *kekHex == "" {
		return fmt.Errorf("missing required flags: --out --pubkey-hex --sk-hex --kek-hex")
	}

	strict := wolfcryptStrict()
	km, cleanup, err := loadKeyWrapper(strict)
	if err != nil {
		return err
	}
	defer cleanup()

	pub, err := hexDecodeStrict(*pubkeyHex)
	if err != nil {
		return fmt.Errorf("pubkey-hex: %w", err)
	}
	kek, err := hexDecodeStrict(*kekHex)
	if err != nil {
		return fmt.Errorf("kek-hex: %w", err)
	}
	if err := mustLen(kek, 32, "kek"); err != nil {
		return err
	}
	sk, err := hexDecodeStrict(*skHex)
	if err != nil {
		return fmt.Errorf("sk-hex: %w", err)
	}
	if len(sk) == 0 || len(sk)%8 != 0 {
		return fmt.Errorf("sk must be non-zero multiple of 8 bytes (AES-KW requirement)")
	}

	// key_id = SHA3-256(pubkey)
	p, pc, err := loadCryptoProvider()
	if err != nil {
		return err
	}
	defer pc()
	keyID, err := p.SHA3_256(pub)
	if err != nil {
		return err
	}

	wrapped, err := keywrap(strict, kek, sk, km)
	if err != nil {
		return err
	}

	ks := KeyStoreV1{
		Version:      "RBKSv1",
		SuiteID:      uint8(*suiteID),
		PubkeyHex:    hex.EncodeToString(pub),
		KeyIDHex:     hex.EncodeToString(keyID[:]),
		WrapAlg:      "AES-256-KW",
		WrappedSKHex: hex.EncodeToString(wrapped),
	}
	b, err := json.Marshal(ks)
	if err != nil {
		return err
	}
	b = append(b, '\n')
	return os.WriteFile(*out, b, 0o600)
}

func readKeystore(path string) (*KeyStoreV1, error) {
	raw, err := os.ReadFile(path) // #nosec G304 -- operator-provided
	if err != nil {
		return nil, err
	}
	var ks KeyStoreV1
	if err := json.Unmarshal(raw, &ks); err != nil {
		return nil, err
	}
	if ks.Version != "RBKSv1" {
		return nil, fmt.Errorf("unsupported keystore version: %q", ks.Version)
	}
	if strings.ToUpper(ks.WrapAlg) != "AES-256-KW" {
		return nil, fmt.Errorf("unsupported wrap_alg: %q", ks.WrapAlg)
	}
	return &ks, nil
}

func cmdKeymgrImportWrapped(argv []string) error {
	fs := flag.NewFlagSet("keymgr import-wrapped", flag.ExitOnError)
	in := fs.String("in", "", "input keystore json path")
	out := fs.String("out", "", "output keystore json path")
	oldKekHex := fs.String("old-kek-hex", "", "old AES-256 KEK (32 bytes hex)")
	newKekHex := fs.String("new-kek-hex", "", "new AES-256 KEK (32 bytes hex)")
	_ = fs.Parse(argv)
	if *in == "" || *out == "" || *oldKekHex == "" || *newKekHex == "" {
		return fmt.Errorf("missing required flags: --in --out --old-kek-hex --new-kek-hex")
	}

	strict := wolfcryptStrict()
	km, cleanup, err := loadKeyWrapper(strict)
	if err != nil {
		return err
	}
	defer cleanup()

	ks, err := readKeystore(*in)
	if err != nil {
		return err
	}

	oldKek, err := hexDecodeStrict(*oldKekHex)
	if err != nil {
		return fmt.Errorf("old-kek-hex: %w", err)
	}
	if err := mustLen(oldKek, 32, "old-kek"); err != nil {
		return err
	}
	newKek, err := hexDecodeStrict(*newKekHex)
	if err != nil {
		return fmt.Errorf("new-kek-hex: %w", err)
	}
	if err := mustLen(newKek, 32, "new-kek"); err != nil {
		return err
	}
	wrapped, err := hexDecodeStrict(ks.WrappedSKHex)
	if err != nil {
		return fmt.Errorf("wrapped_sk_hex: %w", err)
	}

	plain, err := keyunwrap(strict, oldKek, wrapped, km)
	if err != nil {
		return err
	}
	newWrapped, err := keywrap(strict, newKek, plain, km)
	if err != nil {
		return err
	}
	ks.WrappedSKHex = hex.EncodeToString(newWrapped)

	b, err := json.Marshal(ks)
	if err != nil {
		return err
	}
	b = append(b, '\n')
	return os.WriteFile(*out, b, 0o600)
}

func cmdKeymgrVerifyPubkey(argv []string) (string, error) {
	fs := flag.NewFlagSet("keymgr verify-pubkey", flag.ExitOnError)
	in := fs.String("in", "", "input keystore json path")
	expectedKeyIDHex := fs.String("expected-key-id-hex", "", "optional expected key_id hex")
	_ = fs.Parse(argv)
	if *in == "" {
		return "", fmt.Errorf("missing required flag: --in")
	}

	ks, err := readKeystore(*in)
	if err != nil {
		return "", err
	}
	pub, err := hexDecodeStrict(ks.PubkeyHex)
	if err != nil {
		return "", fmt.Errorf("pubkey_hex: %w", err)
	}

	p, cleanup, err := loadCryptoProvider()
	if err != nil {
		return "", err
	}
	defer cleanup()
	keyID, err := p.SHA3_256(pub)
	if err != nil {
		return "", err
	}
	gotHex := hex.EncodeToString(keyID[:])
	if ks.KeyIDHex != "" && !strings.EqualFold(ks.KeyIDHex, gotHex) {
		return "", fmt.Errorf("keystore key_id mismatch: embedded=%s computed=%s", ks.KeyIDHex, gotHex)
	}
	if *expectedKeyIDHex != "" {
		exp := strings.ToLower(strings.TrimPrefix(strings.TrimSpace(*expectedKeyIDHex), "0x"))
		if exp != gotHex {
			return "", fmt.Errorf("expected key_id mismatch: expected=%s computed=%s", exp, gotHex)
		}
	}
	return gotHex, nil
}

func cmdKeymgrMain(argv []string) int {
	if len(argv) < 1 {
		fmt.Fprintln(os.Stderr, "usage: rubin-node keymgr <subcommand> [flags]")
		return 2
	}
	sub := argv[0]
	subargv := argv[1:]

	switch sub {
	case "export-wrapped":
		if err := cmdKeymgrExportWrapped(subargv); err != nil {
			fmt.Fprintln(os.Stderr, "keymgr export-wrapped error:", err)
			return 1
		}
		fmt.Println("OK")
		return 0
	case "import-wrapped":
		if err := cmdKeymgrImportWrapped(subargv); err != nil {
			fmt.Fprintln(os.Stderr, "keymgr import-wrapped error:", err)
			return 1
		}
		fmt.Println("OK")
		return 0
	case "verify-pubkey":
		out, err := cmdKeymgrVerifyPubkey(subargv)
		if err != nil {
			fmt.Fprintln(os.Stderr, "keymgr verify-pubkey error:", err)
			return 1
		}
		fmt.Println(out)
		return 0
	default:
		fmt.Fprintln(os.Stderr, "unknown keymgr subcommand")
		return 2
	}
}

File: clients/rust/crates/rubin-node/src/keymgr.rs
use std::fs;

use crate::aeskw;

trait KeyWrapProvider {
    fn wrap(&self, kek: &[u8], key_in: &[u8]) -> Result<Vec<u8>, String>;
    fn unwrap(&self, kek: &[u8], wrapped: &[u8]) -> Result<Vec<u8>, String>;
}

struct SoftwareKeyWrap;

impl KeyWrapProvider for SoftwareKeyWrap {
    fn wrap(&self, kek: &[u8], key_in: &[u8]) -> Result<Vec<u8>, String> {
        aeskw::aes_key_wrap_rfc3394(kek, key_in)
    }

    fn unwrap(&self, kek: &[u8], wrapped: &[u8]) -> Result<Vec<u8>, String> {
        aeskw::aes_key_unwrap_rfc3394(kek, wrapped)
    }
}

#[cfg(feature = "wolfcrypt-dylib")]
struct ShimKeyWrap {
    p: rubin_crypto::WolfcryptDylibProvider,
}

#[cfg(feature = "wolfcrypt-dylib")]
impl KeyWrapProvider for ShimKeyWrap {
    fn wrap(&self, kek: &[u8], key_in: &[u8]) -> Result<Vec<u8>, String> {
        self.p.key_wrap(kek, key_in)
    }

    fn unwrap(&self, kek: &[u8], wrapped: &[u8]) -> Result<Vec<u8>, String> {
        self.p.key_unwrap(kek, wrapped).map_err(|e| e.to_string())
    }
}

fn parse_hex_flag(args: &[String], flag: &str) -> Result<Vec<u8>, i32> {
    match crate::get_flag(args, flag) {
        Ok(Some(v)) => match rubin_consensus::hex_decode_strict(&v) {
            Ok(b) => Ok(b),
            Err(e) => {
                eprintln!("{flag}: {e}");
                Err(2)
            }
        },
        Ok(None) => {
            eprintln!("missing required flag: {flag}");
            Err(2)
        }
        Err(e) => {
            eprintln!("{e}");
            Err(2)
        }
    }
}

fn parse_opt_hex_flag(args: &[String], flag: &str) -> Result<Option<Vec<u8>>, i32> {
    match crate::get_flag(args, flag) {
        Ok(Some(v)) => match rubin_consensus::hex_decode_strict(&v) {
            Ok(b) => Ok(Some(b)),
            Err(e) => {
                eprintln!("{flag}: {e}");
                Err(2)
            }
        },
        Ok(None) => Ok(None),
        Err(e) => {
            eprintln!("{e}");
            Err(2)
        }
    }
}

fn parse_u8_flag(args: &[String], flag: &str) -> Result<u8, i32> {
    match crate::get_flag(args, flag) {
        Ok(Some(v)) => {
            let s = v.trim();
            let parsed = if let Some(hex) = s.strip_prefix("0x").or_else(|| s.strip_prefix("0X")) {
                u8::from_str_radix(hex, 16)
            } else {
                s.parse::<u8>()
            };
            parsed.map_err(|e| {
                eprintln!("{flag}: {e}");
                2
            })
        }
        Ok(None) => {
            eprintln!("missing required flag: {flag}");
            Err(2)
        }
        Err(e) => {
            eprintln!("{e}");
            Err(2)
        }
    }
}

fn parse_string_flag(args: &[String], flag: &str) -> Result<String, i32> {
    match crate::get_flag(args, flag) {
        Ok(Some(v)) => Ok(v),
        Ok(None) => {
            eprintln!("missing required flag: {flag}");
            Err(2)
        }
        Err(e) => {
            eprintln!("{e}");
            Err(2)
        }
    }
}

fn write_file_0600(path: &str, bytes: &[u8]) -> Result<(), String> {
    use std::io::Write;
    #[cfg(unix)]
    use std::os::unix::fs::OpenOptionsExt;

    let mut opts = std::fs::OpenOptions::new();
    opts.write(true).create(true).truncate(true);
    #[cfg(unix)]
    {
        opts.mode(0o600);
    }
    let mut f = opts.open(path).map_err(|e| format!("write {path}: {e}"))?;
    f.write_all(bytes)
        .map_err(|e| format!("write {path}: {e}"))?;
    Ok(())
}

#[cfg(feature = "wolfcrypt-dylib")]
fn load_key_wrapper(strict: bool) -> Result<Box<dyn KeyWrapProvider>, String> {
    let has_shim_path = std::env::var("RUBIN_WOLFCRYPT_SHIM_PATH")
        .map(|v| !v.trim().is_empty())
        .unwrap_or(false);

    if !has_shim_path {
        if strict {
            return Err("RUBIN_WOLFCRYPT_STRICT=1 requires RUBIN_WOLFCRYPT_SHIM_PATH".into());
        }
        return Ok(Box::new(SoftwareKeyWrap));
    }

    let p = rubin_crypto::WolfcryptDylibProvider::load_from_env()?;
    if p.has_key_management() {
        Ok(Box::new(ShimKeyWrap { p }))
    } else if strict {
        Err("keymgr requires wolfcrypt shim keywrap symbols in strict mode".into())
    } else {
        Ok(Box::new(SoftwareKeyWrap))
    }
}

#[cfg(not(feature = "wolfcrypt-dylib"))]
fn load_key_wrapper(strict: bool) -> Result<Box<dyn KeyWrapProvider>, String> {
    if strict {
        return Err("RUBIN_WOLFCRYPT_STRICT=1 requires feature wolfcrypt-dylib".into());
    }
    Ok(Box::new(SoftwareKeyWrap))
}

fn read_keystore(path: &str) -> Result<serde_json::Map<String, serde_json::Value>, String> {
    let raw = fs::read_to_string(path).map_err(|e| format!("read keystore: {e}"))?;
    let v: serde_json::Value =
        serde_json::from_str(&raw).map_err(|e| format!("keystore json: {e}"))?;
    let obj = v
        .as_object()
        .ok_or_else(|| "keystore json: root must be object".to_string())?;

    let version = obj.get("version").and_then(|v| v.as_str()).unwrap_or("");
    if version != "RBKSv1" {
        return Err(format!("unsupported keystore version: {version:?}"));
    }
    let wrap_alg = obj.get("wrap_alg").and_then(|v| v.as_str()).unwrap_or("");
    if !wrap_alg.eq_ignore_ascii_case("AES-256-KW") {
        return Err(format!("unsupported wrap_alg: {wrap_alg:?}"));
    }
    Ok(obj.clone())
}

fn cmd_keymgr_export_wrapped(args: &[String]) -> i32 {
    let out = match parse_string_flag(args, "--out") {
        Ok(v) => v,
        Err(code) => return code,
    };
    let suite_id = match parse_u8_flag(args, "--suite-id") {
        Ok(v) => v,
        Err(code) => return code,
    };
    let pubkey = match parse_hex_flag(args, "--pubkey-hex") {
        Ok(v) => v,
        Err(code) => return code,
    };
    let sk = match parse_hex_flag(args, "--sk-hex") {
        Ok(v) => v,
        Err(code) => return code,
    };
    let kek = match parse_hex_flag(args, "--kek-hex") {
        Ok(v) => v,
        Err(code) => return code,
    };
    if kek.len() != 32 {
        eprintln!("kek must be 32 bytes (got {})", kek.len());
        return 2;
    }
    if sk.is_empty() || sk.len() % 8 != 0 {
        eprintln!("sk must be non-zero multiple of 8 bytes (AES-KW requirement)");
        return 2;
    }

    let strict = crate::wolfcrypt_strict();
    let km = match load_key_wrapper(strict) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("{e}");
            return 1;
        }
    };

    // key_id = SHA3-256(pubkey)
    let provider = match crate::load_crypto_provider() {
        Ok(p) => p,
        Err(e) => {
            eprintln!("{e}");
            return 1;
        }
    };
    let key_id = match provider.sha3_256(&pubkey) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("{e}");
            return 1;
        }
    };

    let wrapped = match km.wrap(&kek, &sk) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("{e}");
            return 1;
        }
    };

    let mut obj = serde_json::Map::new();
    obj.insert(
        "version".to_string(),
        serde_json::Value::String("RBKSv1".to_string()),
    );
    obj.insert(
        "suite_id".to_string(),
        serde_json::Value::Number(serde_json::Number::from(suite_id)),
    );
    obj.insert(
        "pubkey_hex".to_string(),
        serde_json::Value::String(crate::hex_encode(&pubkey)),
    );
    obj.insert(
        "key_id_hex".to_string(),
        serde_json::Value::String(crate::hex_encode(&key_id)),
    );
    obj.insert(
        "wrap_alg".to_string(),
        serde_json::Value::String("AES-256-KW".to_string()),
    );
    obj.insert(
        "wrapped_sk_hex".to_string(),
        serde_json::Value::String(crate::hex_encode(&wrapped)),
    );

    let mut bytes = serde_json::to_vec(&serde_json::Value::Object(obj))
        .map_err(|e| format!("json: {e}"))
        .unwrap();
    bytes.push(b'\n');
    if let Err(e) = write_file_0600(&out, &bytes) {
        eprintln!("{e}");
        return 1;
    }
    0
}

fn cmd_keymgr_import_wrapped(args: &[String]) -> i32 {
    let in_path = match parse_string_flag(args, "--in") {
        Ok(v) => v,
        Err(code) => return code,
    };
    let out_path = match parse_string_flag(args, "--out") {
        Ok(v) => v,
        Err(code) => return code,
    };
    let old_kek = match parse_hex_flag(args, "--old-kek-hex") {
        Ok(v) => v,
        Err(code) => return code,
    };
    let new_kek = match parse_hex_flag(args, "--new-kek-hex") {
        Ok(v) => v,
        Err(code) => return code,
    };
    if old_kek.len() != 32 {
        eprintln!("old-kek must be 32 bytes (got {})", old_kek.len());
        return 2;
    }
    if new_kek.len() != 32 {
        eprintln!("new-kek must be 32 bytes (got {})", new_kek.len());
        return 2;
    }

    let strict = crate::wolfcrypt_strict();
    let km = match load_key_wrapper(strict) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("{e}");
            return 1;
        }
    };

    let mut ks = match read_keystore(&in_path) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("{e}");
            return 1;
        }
    };
    let wrapped_sk_hex = ks
        .get("wrapped_sk_hex")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    let wrapped = match rubin_consensus::hex_decode_strict(wrapped_sk_hex) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("wrapped_sk_hex: {e}");
            return 1;
        }
    };

    let plain = match km.unwrap(&old_kek, &wrapped) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("{e}");
            return 1;
        }
    };
    let new_wrapped = match km.wrap(&new_kek, &plain) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("{e}");
            return 1;
        }
    };
    ks.insert(
        "wrapped_sk_hex".to_string(),
        serde_json::Value::String(crate::hex_encode(&new_wrapped)),
    );

    let mut bytes = serde_json::to_vec(&serde_json::Value::Object(ks))
        .map_err(|e| format!("json: {e}"))
        .unwrap();
    bytes.push(b'\n');
    if let Err(e) = write_file_0600(&out_path, &bytes) {
        eprintln!("{e}");
        return 1;
    }
    0
}

fn cmd_keymgr_verify_pubkey(args: &[String]) -> i32 {
    let in_path = match parse_string_flag(args, "--in") {
        Ok(v) => v,
        Err(code) => return code,
    };
    let expected = match parse_opt_hex_flag(args, "--expected-key-id-hex") {
        Ok(v) => v,
        Err(code) => return code,
    };

    let ks = match read_keystore(&in_path) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("{e}");
            return 1;
        }
    };
    let pubkey_hex = ks.get("pubkey_hex").and_then(|v| v.as_str()).unwrap_or("");
    let pubkey = match rubin_consensus::hex_decode_strict(pubkey_hex) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("pubkey_hex: {e}");
            return 1;
        }
    };

    let provider = match crate::load_crypto_provider() {
        Ok(p) => p,
        Err(e) => {
            eprintln!("{e}");
            return 1;
        }
    };
    let key_id = match provider.sha3_256(&pubkey) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("{e}");
            return 1;
        }
    };
    let got_hex = crate::hex_encode(&key_id);

    if let Some(embedded) = ks.get("key_id_hex").and_then(|v| v.as_str())
        && !embedded.is_empty()
        && embedded.to_ascii_lowercase() != got_hex
    {
        eprintln!("keystore key_id mismatch: embedded={embedded} computed={got_hex}");
        return 1;
    }
    if let Some(exp) = expected {
        if exp.len() != 32 {
            eprintln!(
                "--expected-key-id-hex must decode to 32 bytes (got {})",
                exp.len()
            );
            return 2;
        }
        let exp_hex = crate::hex_encode(&exp);
        if exp_hex != got_hex {
            eprintln!("expected key_id mismatch: expected={exp_hex} computed={got_hex}");
            return 1;
        }
    }

    println!("{got_hex}");
    0
}

pub fn cmd_keymgr_main(args: &[String]) -> i32 {
    if args.is_empty() {
        eprintln!("usage: rubin-node keymgr <subcommand> [flags]");
        return 2;
    }
    let sub = args[0].as_str();
    let sub_args = &args[1..];
    match sub {
        "export-wrapped" => {
            let rc = cmd_keymgr_export_wrapped(sub_args);
            if rc == 0 {
                println!("OK");
            } else if rc == 1 {
                eprintln!("keymgr export-wrapped error");
            }
            rc
        }
        "import-wrapped" => {
            let rc = cmd_keymgr_import_wrapped(sub_args);
            if rc == 0 {
                println!("OK");
            } else if rc == 1 {
                eprintln!("keymgr import-wrapped error");
            }
            rc
        }
        "verify-pubkey" => cmd_keymgr_verify_pubkey(sub_args),
        _ => {
            eprintln!("unknown keymgr subcommand");
            2
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_keymgr_verify_pubkey_minimal_keystore() {
        let td = tempfile::tempdir().unwrap();
        let ks_path = td.path().join("k.json");

        // Minimal keystore, no wrapped key needed for verify-pubkey.
        // wrapped_sk_hex may be junk, but verify-pubkey must not crash.
        write_file_0600(
            ks_path.to_str().unwrap(),
            br#"{
  "version": "RBKSv1",
  "suite_id": 1,
  "pubkey_hex": "11",
  "key_id_hex": "",
  "wrap_alg": "AES-256-KW",
  "wrapped_sk_hex": "00"
}
"#,
        )
        .unwrap();

        let rc = cmd_keymgr_verify_pubkey(&vec![
            "--in".to_string(),
            ks_path.to_str().unwrap().to_string(),
        ]);
        assert_eq!(rc, 0);
    }
}
